lib/report.js

```diff
@@ -11,55 +11,142 @@ const line = (title, s) => {
+        if (this.testsCount <= 0) this.report();
+      });
+    }
+  }
+
+  report() {
+    // ignore
+  }
+}
+
+class ConciseReporter extends Reporter {
+
+  constructor(options, ...tests) {
+    super(...tests);
+    this.failures = 0;
+    this.successes = 0;
+    this.tests = 0;
+    this.skipped = 0;
+    this.options = Object.assign({ showTodo: false }, this.options);
+    this._testErrorCallback = (test, failMessage) => {
+      // remove one success because 'error' may only be received after end
+      // though the test may have already failed so we need a way to check
+      // if we should do this, i.e. store map of successful tests
+      // this.successes--;
+      test.todo ? this.skipped++ : this.failures++;
+      if (!test.todo || this.options.showTodo) {
+        this.listFailure(test, {
+          message: `Failure after end: ${failMessage}`
+        }, ` Test failed: ${test.caption}`);
+      }
     };
-    metatests.run(show);
-  };
+  }
+
+  _parseTestResults(test) {
+    this.tests++;
+    const testResult = test.success ? 'succeded' : 'failed';
+    let message = ` Test ${testResult}: ${test.caption}\n`;
+    if (test.metadata.filename) message += `  ${test.metadata.filename}`;
+
+    let firstFailure = true;
+    test.results.forEach((res) => {
+      if (res.type === 'subtest') {
+        if (!res.test.success) {
+          const subResult = {
+            type: res.type,
+            caption: res.caption,
+            success: false,
+          };
+          if (firstFailure) {
+            firstFailure = false;
+            console.log('='.repeat(60));
+          }
+          this.listFailure(res.test, subResult, message);
+        }
+        this._parseTestResults(res.test);
+      } else if (!res.success) {
+        test.todo ? this.skipped++ : this.failures++;
+        if (firstFailure) {
+          firstFailure = false;
+          console.log('='.repeat(60));
+        }
+        this.listFailure(test, res, message);
+      } else {
+        this.successes++;
+      }
+    });
+  }
+
+  record(test) {
+    super.record(test);
+    this._parseTestResults(test);
+  }
 
-  metatests.fail = (reason) => {
-    const message = ' Test failed: ' + reason.caption;
-    delete reason.caption;
-    if (reason.stack) reason.stack = renderStack(reason.stack);
-    console.log(list(reason, message) + '\n');
-  };
+  report() {
+    console.log('='.repeat(60));
+    console.log(`Tests finished ${this.tests}.` +
+      `\t[+${this.successes} | -${this.failures} | ~${this.skipped}]\n`);
+    if (this.failures > 0) process.exit(1);
+  }
 
+  listFailure(test, res, message) {
+    if (!test.todo || this.options.showTodo) {
+      if (res.stack) res.stack = renderStack(res.stack);
+      console.log('-'.repeat(60));
+      console.log(list(res, message, ['success']) + '\n');
+    }
+  }
+}
+
+let defaultReporter = new ConciseReporter();
+
+const reporters = {
+  Reporter,
+  ConciseReporter,
+  defaultReporter: () => defaultReporter,
```

Comment:
Just looking at the public interface of the module without diving into implementation, I'd expect this function to always construct a new instance, unless it was named like `getDefaultReporter`. That said, I fail to see any use case for the getter at all. Also, `setDefaultReporter` is actually just `setReporter`, if I understand the code correctly.

Reply:
`getDefaultReporter` will be better. 👍 
It is a getter because default reporter is global for whole `metatests` (defined above this) and you can set it via `setDefaultReporter`. But you can also provide a `Reporter` instance to `metatests.test(caption, func, options, reporter)` and it's called default because if you haven't provided a reporter to test it will use this one (global).